README.md
# clojure-scratch

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-01-01
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-01-01
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojure-scratch/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure-scratch/compare/0.1.0...0.1.1


test/clojure_scratch/core_test.clj
(ns clojure-scratch.core-test
  (:require [clojure.test :refer :all]
            [clojure-scratch.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject clojure-scratch "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [org.clojure/core.async "1.2.603"]
                 [com.taoensso/timbre "4.10.0"]
                 [com.stuartsierra/component "0.3.2"]
                 [clj-http "3.9.0"]
                 [cheshire "5.8.0"]
                 [com.google.apis/google-api-services-bigquery "v2-rev256-1.21.0"]
                 [prismatic/schema "1.1.9"]
                 [liberator "0.15.2"]
                 [ring-middleware-format "0.7.2" :exclusions [cheshire ring/ring-core]]
                 [compojure "1.6.1" :exclusions [ring/ring-core]]
                 [http-kit "2.2.0"]
                 [clojure-csv/clojure-csv "2.0.1"]
                 [ring "1.7.0-RC1" :exclusions [ring/ring-codec clj-time]]
                 [javax.servlet/servlet-api "2.5"]
                 [ring/ring-json "0.4.0" :exclusions [cheshire]]
                 [ring/ring-jetty-adapter "1.7.0-RC1" :exclusions [clj-time ring/ring-core]]
                 [clj-time "0.15.1"]
                 [com.taoensso/carmine "2.18.1" :exclusions [org.clojure/tools.reader com.taoensso/encore]]
                 [org.clojure/data.json "0.2.6"]
                 [org.clojure/tools.namespace "0.2.11"]
                 [org.clojure/core.cache "0.8.2"]]
  :repl-options {:init-ns clojure-scratch.core})


src/compojure/simple_route.clj
(ns compojure.simple-route
  (:require [org.httpkit.server :refer :all]
            [compojure.route :refer :all]
            [compojure.core :refer :all]
            [ring.util.response :as ring-response]
            [clojure.java.io :as io]
            [clojure.pprint :as pprint])
  (:import (java.io ByteArrayInputStream InputStream)))

(defonce server (atom nil))

(defn long-byte-stream [_]
  {:status  200
   :headers {"Content-Type"       "text/csv"
             "Content-Disposition" "attachment; filename=\"long-stream.txt\""}
   :body    (proxy [InputStream] []
              (read
                ([] (rand-int 1000))
                ([^bytes bytes] 1)
                ([^bytes bytes off len] 1)))})
   ;:body    (System/in)})

(defn prettifyRequest [req]
  (str "<pre>" (with-out-str (pprint/pprint req)) "<pre>"))

(defn file [_]
  (ring-response/file-response "README.md"))

(defn stream [_]
  {:status  200
   :headers {"Content-Type"       "application/octet-stream"
             "Content-Disposition" "attachment; filename=\"temp.txt\""}
   :body    (io/input-stream (ByteArrayInputStream. (.getBytes "text")))})

(defroutes
  handler
  (GET "/" [] "hello compojure")
  (GET "/get/:name" [name] (str "Hello " name))
  (GET "/request" [] prettifyRequest)
  (GET "/file" [] file)
  (GET "/stream" [] stream)
  (GET "/long-stream" [] long-byte-stream)
  (not-found "<p>Page not found.</p>"))

(defn stop-server []
  (when-not (nil? @server)
    ;; graceful shutdown: wait 100ms for existing requests to be finished
    ;; :timeout is optional, when no timeout, stop immediately
    (@server :timeout 100)
    (reset! server nil)))

(defn -main [& args]
  ;; The #' is useful when you want to hot-reload code
  ;; You may want to take a look: https://github.com/clojure/tools.namespace
  ;; and http://http-kit.org/migration.html#reload
  (reset! server (run-server #'handler {:port 8081}))
  (println "start server"))

(-main)

src/http_kit/async_response.clj
(ns http-kit.async-response
  (:require [org.httpkit.server :refer :all]
            [clojure.core.async :as async])
  (:import (java.io ByteArrayInputStream InputStream)))

(defn generate-string [len]
  (apply str (repeatedly len #(char (+ (rand 26) 65)))))

(defn sending-loop [channel]
  (loop []
    (send! channel
           {:headers {"Content-Type"        "text/csv"
                      "Content-Disposition" "attachment; filename=\"long-stream.txt\""}
            ;:body    (generate-string (+ (* 50 1024 1024) (rand 10000)))}
            :body
                     (let [stream (ByteArrayInputStream. (.getBytes (generate-string (+ (* 50 1024 1024) (rand 10000))) "UTF-8"))
                           counter (atom 0)
                           process-read (fn [f]
                                          (println "asd")
                                          (if (zero? (mod (swap! counter inc) 10))
                                            (throw (OutOfMemoryError. "test"))
                                            (f)))]
                       (proxy [InputStream] []
                         (read
                           ([] (process-read #(.read stream)))
                           ([^bytes bytes] (process-read #(.read stream bytes)))
                           ([^bytes bytes off len] (process-read #(.read stream bytes off len)))
                           )))}
           false)
    (Thread/sleep 3000)
    (recur)))

(defn handler [req]
  (with-channel req channel

                (on-close channel (fn [_] (println "channel closed")))

                (async/thread-call (fn [] (sending-loop channel)))))


(def server (run-server handler {:port 8291}))

src/async_io/simpe_processor.clj
(ns async-io.simpe-processor
  (:require [clojure.core.async :as async]))

(def input-ch (async/chan 3))
(def terminate-ch (async/chan))

(defn consumer [input-ch terminate-ch]
  (async/thread
    (loop []
      (println "get from channels")
      (let [[v ch] (async/alts!! [input-ch terminate-ch])]
        (when (identical? ch input-ch)
          (println "channel element: " v)
          (Thread/sleep 10000)
          (recur))))))

(defn go-consumer [name input-ch terminate-ch]
  (async/go-loop []
    (println name "consumer get from channel" (Thread/currentThread))
    (let [[v ch] (async/alts! [input-ch terminate-ch])]
      (when (identical? ch input-ch)
        (println name "go-consumer channel element: " v "thread:" (Thread/currentThread))
        (Thread/sleep 10000000)
        (recur)))))


(comment
  (dotimes [i 2] (async/go (async/>! channel i)))
  (dotimes [i 2] (async/>!! input-ch i))

  (async/>!! input-ch -1)
  (async/<!! input-ch)

  (async/alts!! [input-ch (async/timeout 20)])

  (consumer input-ch terminate-ch)

  (go-consumer (str "go-" 100) input-ch terminate-ch)
  (dotimes [i 10] (go-consumer (str "go-" i) input-ch terminate-ch))

  )


src/jetty/jetty_runner.clj
(ns jetty.jetty-runner
  (:require
    [compojure.core :refer [defroutes routes GET]]
    [compojure.route :as route]
    [ring.adapter.jetty :as jetty]
    [clojure.core.async :refer [go chan >! <!] :as a]
    [ring.middleware.reload :refer [wrap-reload]]
    [clj-http.client :as client]
    [ring.core.protocols :as ring-protocols]
    [clojure.java.io :as io])
  (:import (java.io SequenceInputStream ByteArrayInputStream InputStream OutputStream)
           (clojure.lang IPersistentVector)))

(def counter (atom 0))

(defn hello-world []
  (ring.util.response/response
    (ring.util.io/piped-input-stream
      (fn [output-stream]
        (spit output-stream "zxcasd")))))

(extend-protocol ring-protocols/StreamableResponseBody
  IPersistentVector
  (write-body-to-stream [[body] _ ^OutputStream output-stream]
    (do
      (println "extend-protocol")
      (io/copy body output-stream))))

(defn generate-string [len]
  (apply str (repeatedly len #(char (+ (rand 26) 65)))))


(defroutes
   handler
   (GET "/ping" [] (str "pong " (swap! counter inc)))
   (GET "/stream" [] (hello-world))
   (GET "/t/:appId" [app_id :as req]
     (println req)
     req)
   (GET "/huge-stream-with-error" [:as req]
     {:status  200
      :headers {"Content-Type"        "text/plain; charset=utf-8"
                "Content-Disposition" "attachment; filename=myFile.csv"}
      :body
               (let [stream (ByteArrayInputStream. (.getBytes (generate-string 1000000) "UTF-8"))
                     counter (atom 0)
                     process-read (fn [f]
                                    (println "asd" @counter)
                                    (Thread/sleep 1000)
                                    (if (zero? (mod (swap! counter inc) 100))
                                      (throw (OutOfMemoryError. "test"))
                                      (f)))]
                 [(proxy [InputStream] []
                    (read
                      ([] (process-read #(.read stream)))
                      ([^bytes bytes] (process-read #(.read stream bytes)))
                      ([^bytes bytes off len] (process-read #(.read stream bytes off len)))
                      ))])
      #_(SequenceInputStream.
          (ByteArrayInputStream. (.getBytes "asd-1" "UTF-8"))
          (proxy [InputStream] []
            (read [_] (throw (OutOfMemoryError. "test")))))})
   (GET "/sleep" [:as req]
     (println "before sleep")
     (Thread/sleep 1000000)
     (println "after sleep"))
   (GET "/page" []
     {:status 200
      ;:headers {
      ;          "Access-Control-Allow-Origin"  "*"
      ;          "Access-Control-Allow-Methods" "*"
      ;          "Access-Control-Allow-Headers" "*"
      ;          }
      :body   "<!DOCTYPE html>\n<html>\n<body>\n\n<h1>My First JavaScript</h1>\n\n<button type=\"button\" onclick=\"UserAction()\">\nClick me to display Date and Time.</button>\n\n<p id=\"demo\"></p>\n\n<script>\n\tfunction UserAction() {\n    var xhttp = new XMLHttpRequest();\n    xhttp.onreadystatechange = function() {\n         if (this.readyState == 4 && this.status == 200) {\n             alert(this.responseText);\n         }\n    };\n    xhttp.open(\"GET\", \"http://localhost:8154/redirect\");\n    xhttp.setRequestHeader(\"Content-type\", \"application/json\");\n    xhttp.send(\"Your JSON Data Here\");\n}\n</script>\n\n</body>\n</html> \n"})

   (GET "/redirect" []
     ;(:body (client/get "http://192.168.0.101:8080/text")))
     {:status  302
      :headers {
                ;"Access-Control-Allow-Origin"  "*"
                ;"Access-Control-Allow-Methods" "*"
                ;"Access-Control-Allow-Headers" "*"
                "Location" "http://machine:8080/text"}})
   (compojure.route/not-found "no page found"))

(defn -main [& args]
  (jetty/run-jetty
    (wrap-reload #'handler)
    {:port        8154
     :join?       false
     :async?      false
     :min-threads 1
     :max-threads 4}))

